### DISCUSSION TITLE
How to implement very simple multi tenancy?

### DISCUSSION THREAD

**Discussion by GBeushausen** [symfony/symfony#53949] [Q&A]
Hi,
coming from Laravel i'm wondering what is the best way to implement a very simple multi tenancy solution in a new app? I just want that every record out of an entity implementing TenantAwareInterface only reads those records where the tenant_id = the currently loggedin users tenant_id and every record that gets written to those entities automatically adds the tenant_id of the user. I just want to keep it simple and not depend on a 3rd party solution.

In Laravel this is very easy:

`
    protected static function booted(): void
    {
        static::addGlobalScope('tenant', function (Builder $query) {
            if (auth()->user() !== null) {
                $query->where('tenant_id', auth()->user()->tenant_id);
            }
        });

        static::creating(function (self $model) {
            if (auth()->user() !== null) {
                $model->tenant()->associate(auth()->user()->tenant()->value('id'));
            }
        });
    }

`

I'm wondering what is the best way to do it in Symfony? I've experiented with EventListeners (onKernelRequest for example). But those request don't seem to be firing in functional tests.

Speaking of functional tests, i'm wondering how can i login a user in a functional test? I don't want a whole WebTestCase for this. 
In Laravel i can just say 
`        $this->actingAs($user);
`
In Symfony so far i only found a login routine for WebTestCases.

In Laravel my functional tests looks like this:

`    public function testUserOnlySeesHisRecords(): void
    {
        $this->insertNoise(Document::class, 5);
        $user = $this->loginAsUser();
        $expectedDocument = Document::factory()->create([
            'tenant_id' => $user->tenant()->value('id'),
        ]);
        $this->assertModelExists($expectedDocument);
        $documents = Document::all();

        foreach ($documents as $document) {
            $this->assertEquals($expectedDocument->tenant()->value('id'), $document->tenant()->value('id'));
        }
        $this->assertCount(1, $documents);
    }
`
How can i implement something like this in Symfony?

**Comment by MatTheCat** [comment]
Hello,

I don't know how persistence is handled with Laravel but with Symfony you would typically use doctrine, so you could implement a [filter](https://www.doctrine-project.org/projects/doctrine-orm/en/3.0/reference/filters.html). See [the DoctrineBundle's documentation](https://symfony.com/bundles/DoctrineBundle/current/configuration.html#filters-configuration) to use it with Symfony.

> how can i login a user in a functional test? I don't want a whole WebTestCase for this

Did you try [`Security::login()`](https://symfony.com/doc/current/security.html#login-programmatically)?

**Reply by GBeushausen** [reply]
Thanks, that would be great. But how do i access Security::login inside a functional test? I tried booting the Kernel and getting it from the container, but then i received this error message:

`Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException : The "Symfony\Bundle\SecurityBundle\Security" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead.
`

**Reply by MatTheCat** [reply]
Indeed, see https://symfony.com/doc/current/service_container/alias_private.html for an explanation.

You can force the service not to be removed by creating a public alias in the test environment, something like

```yaml
# config/services.yaml
# …
when@test:
    services:
        app.security:
            alias: Symfony\Bundle\SecurityBundle\Security
            public: true
```

Then, your test can access the `app.security` or `Symfony\Bundle\SecurityBundle\Security` service.

**Reply by GBeushausen** [reply]
Thanks. The security component is now available, but it seems you can't login a user in a functional test.
`Symfony\Component\Security\Core\Exception\LogicException : Unable to login without a request context.`

**Reply by MatTheCat** [reply]
Ah my bad. You should probably use the `security.token_storage` service then. It exposes a `setToken` method which you can call with the token your authenticator would have created (`UsernamePasswordToken` e.g.).

This would skip some steps but it may be sufficient in your case.
